File: yql/yql.js
#!/usr/bin/env node
/// yql.js


var sprintf
try {
     sprintf = require('sprintf').sprintf;
} catch (err) {
    console.log('require sprintf module, use \'npm install sprintf\'');
    return;
}

var MAXDAYS = 3;
var YQL = require('yql');

var city_name = 'Taipei';
var conditions = ['item.condition', 'astronomy'];
var condition_str = '*';

//where woeid in (select woeid from geo.places(1) where text='Taipei')
var wherestr = sprintf("where woeid in (select woeid from geo.places(1) where text =\'%s\') and u='c'", city_name);
var yqlstr = sprintf("select %s from weather.forecast ", condition_str);

var query_str = yqlstr + wherestr
do_query(query_str);

function do_query(query_str)
{
    // 'select * from weather.forecast where (location = 94089)'
    // 'select item.condition from weather.forecast where woeid = 2487889'
    var query = new YQL(query_str);

    query.exec(function(err, data) {
        print_out(data);
    });
}

function print_out(data)
{
    console.log("print_out....");
    console.log(data.query.results.channel.astronomy);
    console.log(data.query.results.channel.atmosphere);
    console.log(data.query.results.channel.item.condition);
    console.log("\nforecast in " + MAXDAYS + ' days');
    for (var i = 0; i < MAXDAYS; ++i) {
        console.log(data.query.results.channel.item.forecast[i]);
    }

}

File: yql/yql.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# use yql to query yahoo weather
# to get pretty output:
# ./yql.py|python -m json.tool
#
# from: http://57life.blogspot.tw/2012/12/msn-yahoo.html
# 20070568   台北 (台北市內湖區)

import urllib2, urllib, json

def print_tips():
    print("you may use './yql.py | python -m json.tool' to get pretty format\n")

def query_weather():
    woeid = '20070568'
    city_name = 'Taipei'
    #condition = 'item.condition'
    condition = 'astronomy'
    baseurl = "https://query.yahooapis.com/v1/public/yql?"

    tmp = '''
        select {0} from weather.forecast where woeid in (select woeid
        from geo.places(1) where text =\'{1}\') and u=\'c\'
    '''.format(condition, city_name)
    yql_query = tmp.strip()

    yql_url = baseurl + urllib.urlencode({'q':yql_query}) + "&format=json"
    result = urllib2.urlopen(yql_url).read()
    data = json.loads(result)

    # DO NOT use this
    # need more string processing for further json usage, not good output
    #print data['query']['results']

    # use this:
    print json.dumps(data['query']['results'])


if __name__ == '__main__':
    query_weather()
